tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -0,0 +1,194 @@
+///
+/// Calls TTree::GetListOfFriends and parses its result for the names, aliases
+/// and file names, with different methodologies depending on whether the
+/// parameter is a TTree or a TChain.
+///
+/// \note This function only retrieves information about <b>direct friends</b>
+///       of the input tree. It will not recurse through friends of friends and
+///       does not take into account circular references in the list of friends
+///       of the input tree.
+RFriendInfo GetFriendInfo(const TTree &tree)
+{
+   std::vector<NameAlias> friendNames;
+   std::vector<std::vector<std::string>> friendFileNames;
+
+   // Typically, the correct way to call GetListOfFriends would be `tree.GetTree()->GetListOfFriends()`
+   // (see e.g. the discussion at https://github.com/root-project/root/issues/6741).
+   // However, in this case, in case we are dealing with a TChain we really only care about the TChain's
+   // list of friends (which will need to be rebuilt in each processing task) while friends of the TChain's
+   // internal TTree, if any, will be automatically loaded in each task just like they would be automatically
+   // loaded here if we used tree.GetTree()->GetListOfFriends().
+   const auto *friends = tree.GetListOfFriends();
+   if (!friends)
+      return RFriendInfo();
+
+   for (auto fr : *friends) {
+      // Can't pass fr as const TObject* because TFriendElement::GetTree is not const.
+      // Also, we can't retrieve frTree as const TTree* because of TTree::GetFriendAlias(TTree *) a few lines later
+      auto frTree = static_cast<TFriendElement *>(fr)->GetTree();
+
+      friendFileNames.emplace_back();
+      auto &fileNames = friendFileNames.back();
+
+      // Check if friend tree/chain has an alias
+      const auto *alias_c = tree.GetFriendAlias(frTree);
+      const std::string alias = alias_c != nullptr ? alias_c : "";
+
+      // If the current tree is a TChain
+      if (auto frChain = dynamic_cast<const TChain *>(frTree)) {
+         // Note that each TChainElement returned by TChain::GetListOfFiles has a name
+         // equal to the tree name of this TChain and a title equal to the filename.
+         // Accessing the information like this ensures that we get the correct
+         // filenames and treenames if the treename is given as part of the filename
+         // via chain.AddFile(file.root/myTree) and as well if the tree name is given
+         // in the constructor via TChain(myTree) and a file is added later by chain.AddFile(file.root).
+
+         // Get name of the tree for the current chain
+         const auto *chainFiles = frChain->GetListOfFiles();
+         if (!chainFiles || chainFiles->GetEntries() == 0) {
+            throw std::runtime_error("A TChain in the list of friends does not contain any file. "
+                                     "Friends with no associated files are not supported.");
+         }
+         const auto *realName = chainFiles->First()->GetName();
```

Comment:
It feels odd that only the 'first' name matter ....

Reply:
This is done specifically for cases such as
```
root [0] TChain ch{};
root [1] ch.Add("Run2012BC_DoubleMuParked_Muons.root/Events");
root [2] ch.GetName()
(const char *) ""
root [3] const auto *chainFiles = ch.GetListOfFiles();
root [4] chainFiles->First()->GetName()
(const char *) "Events"
root [5] 
```
As also described in the commented lines 107-112

Reply:
Ok but then what about:
```
TChain ch{};
 ch.Add("Run2012BC_DoubleMuParked_Muons1.root/Events1");
 ch.Add("Run2012BC_DoubleMuParked_Muons2.root/Events2");
 ch.Add("Run2012BC_DoubleMuParked_Muons3.root/Events3");
```

Reply:
Right, the function `GetTreeFullPaths` would return a vector with {"Events1","Events2","Events3"}, but it's kind of implicit in the logic of this function `GetFriendInfo` that, on top of retrieving information only about direct friends, there is a 1:1 relation between any TFriendElement's tree(chain) and the original tree(chain), i.e. the situation of an original tree with a single chain as friend but this friend having multiple different trees inside it is un-treated ( and I'm deeply hoping it doesn't happen often :cold_sweat: ). After all, the type of the return value of this function already states that quite clearly:
```
using NameAlias = std::pair<std::string, std::string>; ///< A pair of name and alias of a TTree's friend tree.
struct RFriendInfo {
   std::vector<NameAlias> fFriendNames;                    ///< Pairs of names and aliases of friend trees/chains
```

Reply:
That being said, I wonder myself as to how that kind of usecase would be treated by RDataFrame or if anybody has ever done a TTreeProcessorMT application with chain friends having multiple different trees inside them. Probably Enrico can weigh better on this one

Reply:
I tried to understand better the usecase with this gist https://gist.github.com/vepadulano/6aa86c34f58ca2008f2f175202e205c1 
I create a ttree and then a tchain with multiple ttrees inside of it. The tchain is then added as a friend to the initial ttree. Finally, I draw a graph of the "x" variable of the original tree against the "x" variable of one of the friend trees.
With `TTree::Draw` it runs fine, but RDF crashes instead.
I'm really not sure if I did something wrong but I wanted to share this so maybe I can understand better what we could be dealing with

Reply:
uhm what is RDF supposed to do there? if I understand correctly `treefriend1.x` only refers to the entries in the _first_ tree of the friend chain, so it refers to N entries while the main tree has M > N entries?

I don't think this case is supported and the error message is reasonable. Also we have not had a single user complain about lack of support for differently-named trees in a friend chain in 4 years: feel free to open an issue, but it should not block this PR.

Reply:
Each of the trees is built with exactly N entries and exactly the same values, so `treeparent.x` is exactly equal to `treefriend1.x`. The output of the TTree::Draw part of the gist is

![friendchain_multiple_friendtrees_ttree_draw](https://user-images.githubusercontent.com/15638895/113996544-92f46900-9857-11eb-8db5-270713359ff4.png)

I agree that this is outside of the scope of this PR though, will probably merge it and create an issue only for this

Reply:
If I read the gist correctly you have one main tree with 10k entries with one friend TChain with 5 friends, each with 10k entries (i.e. the friend chain has 50k entries), is that correct?

If so: friends with a different number of entries than the main trees are not supported, and I am not sure in general there is a good branch name that can be used for all branches called "x" in all friends, if the friend trees have different names and the TChain does not have a name.

Reply:
I'm full of doubts about this use case too, but just for the sake of understanding it, I thought the gist (which is admittedly contrived) represented the following case:
| treeparent.x | treefriend1.x | treefriend2.x | treefriend3.x | treefriend4.x | treefriend5.x |
|--------------|---------------|---------------|---------------|---------------|---------------|
| 1            | 1             | 1             | 1             | 1             | 1             |
| 2            | 2             | 2             | 2             | 2             | 2             |
| 3            | 3             | 3             | 3             | 3             | 3             |
| 4            | 4             | 4             | 4             | 4             | 4             |
| 5            | 5             | 5             | 5             | 5             | 5             |

Which would be perfectly fine (I guess!) If one did it like
```
TTree parent{"treeparent","treeparent"};
TTree friend1{"treefriend1","treefriend1"};
TTree friend2{"treefriend2","treefriend2"};
...

parent.AddFriend(friend1);
parent.AddFriend(friend2);
...

parent.Draw("x:treefriend1.x");
```

But instead in the gist the various friend trees (with exactly the same dimensionality as the parent tree) are stored in a TChain instead

Reply:
I updated the gist with this other usecase of separate friend TTrees, and indeed `parenttree->Draw("x:treefriend1.x")` gives the same result as with the TChain scenario:
![separate_multiple_friendtrees_ttree_draw](https://user-images.githubusercontent.com/15638895/114006445-bb349580-9860-11eb-90f9-c76e1a4ebf34.png)

I went one step further and tried  `parenttree->Draw("x:treefriend1.x:treefriend2.x")` which outputs
![separate_multiple_friendtrees_ttree_draw](https://user-images.githubusercontent.com/15638895/114006587-db645480-9860-11eb-9799-6bc7acd8f717.png)

And from this plots it looks to me that all the trees and the "x" columns have same dimensionality and values

Reply:
And I also included this approach with RDF and `Graph` in the gist, indeed producing the correct result
![separate_multiple_friendtrees_rdf](https://user-images.githubusercontent.com/15638895/114007467-aad0ea80-9861-11eb-9b9e-187ca4c7a91c.png)

I also copy the function with RDF here:
```
void with_separate_ttrees_rdf(){
    TFile parentfile{"treeparent.root", "read"};
    TFile friendfile1{"treefriend1.root", "read"};
    TFile friendfile2{"treefriend2.root", "read"};
    TFile friendfile3{"treefriend3.root", "read"};
    TFile friendfile4{"treefriend4.root", "read"};
    TFile friendfile5{"treefriend5.root", "read"};


    std::unique_ptr<TTree> parenttree{parentfile.Get<TTree>("treeparent")};
    std::unique_ptr<TTree> friendtree1{friendfile1.Get<TTree>("treefriend1")};
    std::unique_ptr<TTree> friendtree2{friendfile2.Get<TTree>("treefriend2")};
    std::unique_ptr<TTree> friendtree3{friendfile3.Get<TTree>("treefriend3")};
    std::unique_ptr<TTree> friendtree4{friendfile4.Get<TTree>("treefriend4")};
    std::unique_ptr<TTree> friendtree5{friendfile5.Get<TTree>("treefriend5")};

    parenttree->AddFriend(friendtree1.get());
    parenttree->AddFriend(friendtree2.get());
    parenttree->AddFriend(friendtree3.get());
    parenttree->AddFriend(friendtree4.get());
    parenttree->AddFriend(friendtree5.get());

    ROOT::RDataFrame rdf{*parenttree};
    TCanvas c{};
    auto rdfgraph = rdf.Graph("x", "treefriend1.x");
    rdfgraph->DrawClone();
    c.SaveAs("separate_multiple_friendtrees_rdf.png");
}
```

Reply:
now that you updated the gist my comment above does not apply anymore. RDF should support what currently is in the gist.

Reply:
Alright! So for this PR I will include the last suggestion by Philippe, then extend the documentation of this function with a note that this usecase is currently not supported. Will open an issue to tackle this within RDF in another PR :+1: